### ISSUE TITLE
Migration issue: incompletely cleared indexes can mess up migration

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5158]
Reported on the mailing list by Lucas Miranda:
https://groups.google.com/d/msg/rethinkdb/cynZc-_Z-DE/wDhlGwCDBwAJ

He did an `index-rebuild` on 2.0.5, and then upgraded to RethinkDB 2.2.1.
RethinkDB 2.2.1 started migrating some metadata, and then failed as follows:
```C++
Error in src/rdb_protocol/datum.cc at line 1153:
Unreachable code:
Backtrace:
Fri Nov 27 16:54:30 2015

1 [0xa57a20]: backtrace_t::backtrace_t() at ??:?
2 [0xa57db3]: format_backtrace(bool) at ??:?
3 [0xc99d65]: report_fatal_error(char const*, int, char const*, ...) at ??:?
4 [0x8c5960]: ql::parse_secondary(std::string const&) at ??:?
5 [0x8c5b02]: ql::datum_t::extract_primary(store_key_t const&) at ??:?
6 [0x7f3c2b]: clear_sindex_traversal_cb_t::handle_pair(scoped_key_value_t&&, signal_t*) at ??:?
7 [0x9ef745]: btree_depth_first_traversal(counted_t<counted_buf_lock_and_read_t>, key_range_t const&, depth_first_traversal_callback_t*, access_t, direction_t, btree_key_t const*, btree_key_t const*, signal_t*) at ??:?
8 [0x9ef4ae]: btree_depth_first_traversal(counted_t<counted_buf_lock_and_read_t>, key_range_t const&, depth_first_traversal_callback_t*, access_t, direction_t, btree_key_t const*, btree_key_t const*, signal_t*) at ??:?
9 [0x9ef4ae]: btree_depth_first_traversal(counted_t<counted_buf_lock_and_read_t>, key_range_t const&, depth_first_traversal_callback_t*, access_t, direction_t, btree_key_t const*, btree_key_t const*, signal_t*) at ??:?
10 [0x9ef4ae]: btree_depth_first_traversal(counted_t<counted_buf_lock_and_read_t>, key_range_t const&, depth_first_traversal_callback_t*, access_t, direction_t, btree_key_t const*, btree_key_t const*, signal_t*) at ??:?
11 [0x9efcf7]: btree_depth_first_traversal(superblock_t*, key_range_t const&, depth_first_traversal_callback_t*, access_t, direction_t, release_superblock_t, signal_t*) at ??:?
12 [0x7eccb0]: store_t::clear_sindex_data(uuid_u, value_sizer_t*, deletion_context_t const*, key_range_t const&, signal_t*) at ??:?
13 [0x934dc8]: /usr/bin/rethinkdb() [0x934dc8] at 0x934dc8 ()
14 [0x936118]: /usr/bin/rethinkdb() [0x936118] at 0x936118 ()
15 [0x97c548]: coro_t::run() at ??:?
Exiting.
```

From line 12 in the backtrace we can see that this was caused by `clear_sindex_data` being called on an old (pre 1.16) index. I think this can happen if RethinkDB 2.2.1 is started before the old index was fully cleared after the `index-rebuild`.

Unfortunately we don't currently expose the index clearing process in ReQL, so it's hard to check for this condition in `index-rebuild`. We should think about changing that, so that `index-rebuild` can wait for the clearing to complete.

More importantly, it looks like the pre-check we do to detect any unsupported indexes before starting to migrate data doesn't catch deleted indexes with an unsupported version.
If that happens (like in this case), the database will be left in a partially migrated state, essentially making all data inaccessible.
We should fix this asap.

Assigning @Tryneus .

**Comment by Tryneus** [comment]
After visual inspection with @danielmewes, it appears this is related to some sindex post-construction changes introduced in v2.2.  Since he is more familiar with this code, he offered to take this issue.  Reassigning.

**Comment by danielmewes** [comment]
A fix is in branch `daniel_5158` and internal code review 3375.

**Comment by danielmewes** [comment]
Fixed in `v2.2.x` 54a3433 and `next` 156285d.

We will ship the bug fix with RethinkDB 2.2.2.
